### ISSUE TITLE
In afterUpdate associations are null

### ISSUE THREAD

**Issue by tobalsgithub** [balderdashy/waterline#803]
I'm trying to validate a field in a record based on another field. For example, I've got an Application model that looks something like this:

```
module.exports = {

  attributes: {
    organization: {model: 'Organization'},
    environmentFlag: {
      type: 'string',
      enum: ['Staging', 'Production'],
      defaultsTo: 'Staging'
    },
    callbackUrl: {type: 'string'},    
  },

  afterUpdate: function (values, cb) {
    if (values.environmentFlag === 'Production' && values.callbackUrl.substring(0,5).toLowerCase() !== 'https') {
      Application.update(values.id, {
        environmentFlag: 'Staging'
      }, function (err, apps) {
        return cb('Production applications must use HTTPS for their callback URL');
      });      
    } else {
      cb();
    }
  },
};
```

A Staging Application should only be allowed to have a callbackUrl that uses https. 

Using the beforeUpdate lifecycle callback doesn't work, because not all values will necessarily be present, and the ID won't necessarily be preset either, so I can't load the record. So I went with afterUpdate.

However, when using afterUpdate, I noticed that the association field `organization` is set to null in the lifecycle callback, despite it not being one of the updated fields and despite it being set previously. 

If I do nothing in the lifecycle callback, the application is returned from Application.update with `organization` set appropriately.  However, if I do what's in my code above, even though I'm not updating the `organization` field, null gets persisted to the database. 

Any ideas what's going on? Or better ways to validate a field in a record that depends on another field?

**Comment by tobalsgithub** [comment]
To make things simpler, it looks like any time you return an error in the afterUpdate callback, associations are lost. I should mention we're using the sails-postgres adapter.

**Comment by tobalsgithub** [comment]
sails version 0.10.5
sails-postgresql version 0.10.9
waterline version 0.10.14

**Comment by dmarcelino** [comment]
@tobalsgithub, are you still experiencing this? Have you found a solution? Thanks

**Comment by tobalsgithub** [comment]
@dmarcelino, we implemented a totally different code flow to get around this, doing the validation before updating the record. It's not as streamlined as we would like, since it's not just built into the Model, but it was the only thing we got to work. 

I haven't tried since updating to sails version 0.11.

**Comment by devinivy** [comment]
I wonder if this is another victim of nested updates.  I expect a combination of #646 and https://github.com/balderdashy/waterline/issues/964#issuecomment-95626676.

**Comment by alin-roman-zitec** [comment]
Sorry if I'm late to the party, but this is how I would approach it:

    module.exports = {

        types: {
            requireHttps: function(url) {
                return url.substring(0, 5).toLowerCase() === 'https';
            }
        },

        attributes: {
            organization: {model: 'Organization'},
            environmentFlag: {
                type: 'string',
                enum: ['Staging', 'Production'],
                defaultsTo: 'Staging'
            },
            callbackUrl: {
                type: 'string',
                requireHttps: function() {
                    return this.environmentFlag === 'Production';
                }
            }
        }
    };

**Comment by dmarcelino** [comment]
@tobalsgithub, @devinivy, I don't think waterline provides associations in an `afterUpdate` callback since it uses the outputs of the `adapter.update()` call which has no concept of association...

**Comment by tobalsgithub** [comment]
Yeah, that makes some sense. Still doesn't seem to make sense that if we do an `Application.update` in the `afterUpdate` callback, that all associations get set to null.

**Comment by dmarcelino** [comment]
Agreed, with the code in your example:
```javascript
Application.update(values.id, {
        environmentFlag: 'Staging'
      }
```

the `organisation` association should not be touched. Intriguing...

**Comment by sailsbot** [comment]
Thanks for posting, @tobalsgithub.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
